### DISCUSSION TITLE
AddBreadCrumbs to Native Exception

### DISCUSSION THREAD

**Discussion by sleeper-jason** [getsentry/sentry-react-native#2762] [Q&A]
We use beforeSend to add new breadcrumbs and information before sending an event.
This code doesn't seem to run when a native exception fires.
I've tried using https://github.com/a7ul/react-native-exception-handler to adjust the event as well but that doesn't end up changing the sentry event.

What is the recommended way to add data to a native exception?
Thanks

**Comment by krystofwoldrich** [comment]
As you said the `beforeSend` from the JS layer doesn't get called when a native exception happens.

Breadcrumbs added by `Sentry.addBreadcrumb(breadcrumb)` are synced down to the native layer and included in the native events.

But I don't know if that would work for you or if you need to include the information at the moment of the native exception. Then you would need to initialize the native SDKs manually as described [in the docs](https://docs.sentry.io/platforms/react-native/manual-setup/native-init/).

**Comment by sleeper-jason** [comment]
I wasn't sure if beforeSend get's called. Good to know I shouldn't expect that thanks!

I tried a different route which works *sometimes*. I add breadcrumbs using react-native-exception-handler. these breadcrumbs show up at random spots in the breadcrumb list as the timestamp is often one of the earliest. 
I also don't see this exception being reported. Will sentry not report after a exception fires? It seems that today sentry only reports after app opens because I see breadcrumbs from the next app start 

```js
const nativeExceptionHandler = (exceptionString) => {
  addAppExtrasCrumbs();
 Sentry.captureException(new Error(exceptionString));
};

setNativeExceptionHandler(nativeExceptionHandler, false /* forceQuitApp */, true /* execute default handler */);
```

**Reply by krystofwoldrich** [reply]
I've not tested `react-native-exception-handler` with `sentry-react-native`, but I think this is what's happening.

The native crash captured by the native sentry error handler is saved on the disk and sent after reopening the app. The error created in the JS layer by the `nativeExceptionHandler` is not marked as fatal, so Sentry tries to send it right away, but as the app is crashing the error won't make it to the server.
